File: clients/java/src/test/java/com/retailcortex/modenv/IntegrationTest.java
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.retailcortex.modenv;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import static org.junit.jupiter.api.Assertions.*;

public class IntegrationTest {

    public static class IntegrationConfig {
        public String integrationVal;
        public int port;
    }

    @BeforeEach
    public void setUp() {
        EnvManager.getInstance().restore();
    }

    @AfterEach
    public void tearDown() {
        EnvManager.getInstance().restore();
    }

    private static Path getTestConfigsDir() {
        Path curr = Paths.get("").toAbsolutePath();
        while (curr != null) {
            Path candidate = curr.resolve("test/configs");
            if (Files.isDirectory(candidate)) {
                return candidate;
            }
            curr = curr.getParent();
        }
        return Paths.get("test/configs").toAbsolutePath();
    }

    @Test
    public void testIntegrationEnvManager() {
        EnvManager manager = new EnvManager();
        manager.unset("INTEGRATION_KEY_1");
        manager.unset("INTEGRATION_KEY_2");

        manager.set("INTEGRATION_KEY_1", "val1");
        manager.set("INTEGRATION_KEY_2", "val2");

        assertEquals("val1", manager.get("INTEGRATION_KEY_1"));
        assertEquals("val2", manager.get("INTEGRATION_KEY_2"));

        manager.restore();

        assertFalse(manager.lookup("INTEGRATION_KEY_1").exists());
        assertFalse(manager.lookup("INTEGRATION_KEY_2").exists());
    }

    @Test
    public void testIntegrationLoad() throws IOException {
        Path dir = getTestConfigsDir().resolve("integration");
        EnvManager.getInstance().set("MODENV_PREFIX", dir.toString());

        IntegrationConfig cfg = new IntegrationConfig();
        IntegrationConfig clone = Modenv.load(cfg);

        assertEquals("base", cfg.integrationVal);
        assertEquals(2000, cfg.port);

        assertEquals("base", clone.integrationVal);
        assertEquals(2000, clone.port);
    }
}

File: clients/typescript/test/integration.test.ts
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import * as path from "node:path";
import * as fs from "node:fs";
import { EnvManager, load } from "../src/index.js";

const getTestConfigsDir = (): string => {
  let curr = process.cwd();
  while (curr) {
    const candidate = path.join(curr, "test", "configs");
    if (fs.existsSync(candidate) && fs.statSync(candidate).isDirectory()) {
      return candidate;
    }
    const parent = path.dirname(curr);
    if (parent === curr) break;
    curr = parent;
  }
  return path.resolve("test/configs");
};

describe("modenv TypeScript Integration", () => {
  it("test integration env manager", () => {
    const manager = new EnvManager();
    delete process.env.INTEGRATION_KEY_1;
    delete process.env.INTEGRATION_KEY_2;

    manager.set("INTEGRATION_KEY_1", "val1");
    manager.set("INTEGRATION_KEY_2", "val2");

    assert.equal(process.env.INTEGRATION_KEY_1, "val1");
    assert.equal(process.env.INTEGRATION_KEY_2, "val2");

    manager.restore();

    assert.equal("INTEGRATION_KEY_1" in process.env, false);
    assert.equal("INTEGRATION_KEY_2" in process.env, false);
  });

  it("test integration load", () => {
    const dir = path.join(getTestConfigsDir(), "integration");
    process.env.MODENV_PREFIX = dir;

    interface IntegrationConfig {
      integration_val: string;
      port: number;
    }

    const cfg: IntegrationConfig = {
      integration_val: "",
      port: 0,
    };

    const clone = load<IntegrationConfig>(cfg);

    assert.equal(cfg.integration_val, "base");
    assert.equal(cfg.port, 2000);

    assert.equal(clone.integration_val, "base");
    assert.equal(clone.port, 2000);
  });
});
